runtime/src/snapshot_controller.rs

```diff
@@ -135,3 +135,101 @@ impl SnapshotController {
             })
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use {
+        super::*, crate::accounts_background_service::SnapshotRequestKind,
+        agave_snapshots::snapshot_config::SnapshotConfig, crossbeam_channel::unbounded,
+        solana_genesis_config::create_genesis_config, solana_pubkey::Pubkey, std::sync::Arc,
+        test_case::test_case,
+    };
+
+    fn create_banks(num_banks: u64) -> Vec<Arc<Bank>> {
+        let mut banks = vec![];
+        let (genesis_config, _) = create_genesis_config(1_000_000);
+        let mut parent_bank = Arc::new(Bank::new_for_tests(&genesis_config));
+        banks.push(parent_bank.clone());
+
+        for _i in 1..=num_banks {
+            let new_bank = Arc::new(Bank::new_from_parent(
+                parent_bank.clone(),
+                &Pubkey::default(),
+                parent_bank.slot() + 1,
+            ));
+            parent_bank = new_bank;
+            banks.push(parent_bank.clone());
+        }
+
+        banks
+    }
+    #[test_case(SnapshotInterval::Disabled, SnapshotInterval::Disabled,
+        50, None, None; "Snapshots Disabled")]
+    #[test_case(SnapshotInterval::Slots(10.try_into().unwrap()), SnapshotInterval::Slots(5.try_into().unwrap()),
+        4, None, None; "No snapshot triggered")]
+    #[test_case(SnapshotInterval::Slots(5.try_into().unwrap()), SnapshotInterval::Disabled,
+        8, Some(5), Some(SnapshotRequestKind::FullSnapshot); "Full without Incremental")]
+    #[test_case(SnapshotInterval::Slots(10.try_into().unwrap()), SnapshotInterval::Slots(10.try_into().unwrap()),
+        10, Some(10), Some(SnapshotRequestKind::FullSnapshot); "Full and Incremental on same slot, full is selected")]
+    #[test_case(SnapshotInterval::Slots(10.try_into().unwrap()), SnapshotInterval::Slots(5.try_into().unwrap()),
+        15, Some(15), Some(SnapshotRequestKind::IncrementalSnapshot); "Newer incremental picked over older Full")]
+    #[test_case(SnapshotInterval::Disabled, SnapshotInterval::Slots(5.try_into().unwrap()),
+        5, Some(5), Some(SnapshotRequestKind::IncrementalSnapshot); "Incremental without Full")]
```

Comment:
This is an invalid snapshot configuration, but I see that's not something enforced by SnapshotController. Nothing to change in this PR, just noting.